contributors/design-proposals/container-device-interface.md

```diff
@@ -0,0 +1,138 @@
+
+## Description
+
+CDI is modelled after CNI in that it consist of an API between the Kubelet and CDI binaries that transports instructions through environment variables and configuration and data through JSON chunks transmitted through stdin/stdout.  As in CNI, which binaries to call and their configuration is described through JSON configuration files located in `/etc/cdi/<device-type>.d/<binary-name>.conf`.  The principal commands are `ADD` and `DEL`, which allocate a device to a container or return it to the node's pool.  All commands expect a `CDI_VERSION` environment variable, to ensure the version used by the Kubelet (environment variable), the CDI config (JSON chunk on stdin) and the binary match.
+
+### ADD
+
+`CDI_COMMAND=ADD` allocates a device to a container.  The Kubelet communicates the requested amount of resources through a `CDI_REQUEST=<resource-subtype>:<amount>,...` environment variable.  The CDI binary creates or picks suitable devices, does any potentially necessary setup, stores the association `host device node <> container` and returns the host path of the device node.  The Kubelet will then pass this information as a `Device` to CRI.  Failure of this command results in a non-zero exit status and a JSON encoded error message on stdout.
+
+### DEL
+
+`CDI_COMMAND=DEL` returns devices allocated to a container to the pool.  The CDI binary retrieves the list of device nodes allocated to this container from its own storage, runs any potentially necessary tear down steps and clears the device node / container association from its storage.  This command can not fail, i.e. it always returns a zero exit status, but might produce an error message on stdout, suitable for logging the incident.
+
+## Rationale
+
+We propose out-sourcing the management of device allocations to external binaries, to easily allow to support different device types and provide the possibility for future extensions e.g. using virtualisation techniques.  (In our case, e.g. PCIe SR-IOV could be used to dynamically allocate a certain amount of timeslices or VRAM.)  We propose an API similar to CNI, since its design is already well tested within Kubernetes and capable of this task.
+
+## Usage examples
+
+### GPU / DRM
+
+```
+# cat /etc/cdi/gpu.d/drm.conf
+{
+  "cdiVersion": "0.0.1",
+  "name": "drm-gpus",
+  "type": "drm",
```

Comment:
Thanks for the suggestion! The next version will have this changed. I chose `plugin` instead of `plugin-name` to prevent confusion with the `name` field.